### PR TITLE
support predefined learning rate policy #16

### PR DESCRIPTION
**Pull request by wenyushi**
The push will support predefined learning rate policy, such as cyclic learning rate, reduce learning rate on plateau. 
The push will affect the following files:
1. dlpy/lr_scheduler.py: new module to support predefined learning rate policy
2. dlpy/model.py: adding lr_schduler in the Optimize
3, dlpy/tests/datasources/sample_syntax_for_test.json: fix one benchmark
4. dlpy/tests/test_lr_scheduler.py: unit test covering the lr_scheduler.
5. dlpy/utils.py: move DLPyDict to the file.
6. examples/learning_rate_policy/Define_Learning_Rate_Policy.ipynb: an example showing how to use lr_scheduler

### CODE REVIEW

File: dlpy/model.py [hunk 1]
```diff
@@ -662,9 +662,6 @@ def evaluate_object_detection(self, ground_truth, coord_type, detection_data=Non
         if coord_type.lower() not in ['yolo', 'coco']:
             raise ValueError('coord_type, {}, is not supported'.format(coord_type))
 
-        self.conn.update(table=dict(name = self.model_name, where='_DLChrVal_ eq "iouThreshold"'),
```
Comment (ethem-kinginthenorth) [inline dlpy/model.py]: why we are removing this? is this related to this push?

File: dlpy/model.py [hunk 2]
```diff
@@ -2012,83 +2029,78 @@ def display(self, layer_id, filter_id=None):
         plt.show()
 
 
-class DLPyDict(collections.MutableMapping):
```
Comment (ethem-kinginthenorth) [inline dlpy/model.py]: also why removing this?

File: dlpy/model.py [hunk 3]
```diff
@@ -2012,83 +2029,78 @@ def display(self, layer_id, filter_id=None):
...
-
-    def __keytransform__(self, key):
-        return key.lower().replace("_", "")
-
-    def __str__(self):
-        return str(self.__dict__)
-
-
-class Solver(DLPyDict):
+class _Solver(DLPyDict):
```
Comment (ethem-kinginthenorth) [inline dlpy/model.py]: I think we should do something like this _Solver expands Solver so that we don't really remove the Solver and it will be backward compatible.

File: dlpy/model.py (hunk 1 above)
Comment (wenyushi) [reply dlpy/model.py]: It is not related to this push. I just combined some fixes with the push.

File: dlpy/model.py (hunk 2 above)
Comment (wenyushi) [reply dlpy/model.py]: I think the class is better to move to utils.py

File: dlpy/model.py (hunk 1 above)
Comment (ethem-kinginthenorth) [reply dlpy/model.py]: ok

File: dlpy/model.py (hunk 3 above)
Comment (wenyushi) [reply dlpy/model.py]: The reason I want to make it an abstract class is that Solver is not used to be instantiated. In its child class, we call set_method to add the property. Although we can directly create a Solver object since the method default as Vanilla, Solver will be exactly identical to VanilaSolver. In coding design perspective, I don't think we should keep Solver and instantiate Solver is not right.

Comment (ethem-kinginthenorth) [reply dlpy/model.py]: we cannot remove Solver, any notebook so far would not work with this removal. Please find another way.

**Comment by wenyushi** [comment]
After discussions with Ethem, we would like to do more tests on the Solver change. I just created another branch in my fork to test the push. Here is the branch: https://github.com/wenyushi/python-dlpy/tree/solver2solve.

**Comment by wenyushi** [comment]
Ethem has already merged the change. Close the pull request.
